File: android/app/src/main/java/ai/bluepond/wellness/data/model/Participant.java
package ai.bluepond.wellness.data.model;

import com.google.gson.annotations.SerializedName;

public class Participant {

    @SerializedName("id")
    private String id;

    @SerializedName("auth_user_id")
    private String authUserId;

    @SerializedName("display_name")
    private String displayName;

    @SerializedName("employee_id")
    private String employeeId;

    @SerializedName("email")
    private String email;

    @SerializedName("department")
    private String department;

    @SerializedName("team")
    private String team;

    @SerializedName("height_cm")
    private double heightCm;

    @SerializedName("weight_kg")
    private double weightKg;

    @SerializedName("bmi")
    private double bmi;

    @SerializedName("blood_group")
    private String bloodGroup;

    @SerializedName("shift_type")
    private String shiftType;

    @SerializedName("shift_start_time")
    private String shiftStartTime;

    @SerializedName("shift_end_time")
    private String shiftEndTime;

    @SerializedName("reminder_window_1")
    private String reminderWindow1;

    @SerializedName("reminder_window_2")
    private String reminderWindow2;

    @SerializedName("reminder_frequency")
    private String reminderFrequency;

    @SerializedName("role")
    private String role;

    @SerializedName("status")
    private String status;

    @SerializedName("consent_accepted")
    private boolean consentAccepted;

    @SerializedName("consent_date")
    private String consentDate;

    @SerializedName("onboarding_complete")
    private boolean onboardingComplete;

    @SerializedName("current_streak")
    private int currentStreak;

    @SerializedName("longest_streak")
    private int longestStreak;

    @SerializedName("created_at")
    private String createdAt;

    @SerializedName("updated_at")
    private String updatedAt;

    public Participant() {}

    // ── Getters ──────────────────────────────────────────────────────────────────

    public String getId() { return id; }
    public String getAuthUserId() { return authUserId; }
    public String getDisplayName() { return displayName; }
    public String getEmployeeId() { return employeeId; }
    public String getEmail() { return email; }
    public String getDepartment() { return department; }
    public String getTeam() { return team; }
    public double getHeightCm() { return heightCm; }
    public double getWeightKg() { return weightKg; }
    public double getBmi() { return bmi; }
    public String getBloodGroup() { return bloodGroup; }
    public String getShiftType() { return shiftType; }
    public String getShiftStartTime() { return shiftStartTime; }
    public String getShiftEndTime() { return shiftEndTime; }
    public String getReminderWindow1() { return reminderWindow1; }
    public String getReminderWindow2() { return reminderWindow2; }
    public String getReminderFrequency() { return reminderFrequency; }
    public String getRole() { return role; }
    public String getStatus() { return status; }
    public boolean isConsentAccepted() { return consentAccepted; }
    public String getConsentDate() { return consentDate; }
    public boolean isOnboardingComplete() { return onboardingComplete; }
    public int getCurrentStreak() { return currentStreak; }
    public int getLongestStreak() { return longestStreak; }
    public String getCreatedAt() { return createdAt; }
    public String getUpdatedAt() { return updatedAt; }

    // ── Setters ──────────────────────────────────────────────────────────────────

    public void setId(String id) { this.id = id; }
    public void setAuthUserId(String authUserId) { this.authUserId = authUserId; }
    public void setDisplayName(String displayName) { this.displayName = displayName; }
    public void setEmployeeId(String employeeId) { this.employeeId = employeeId; }
    public void setEmail(String email) { this.email = email; }
    public void setDepartment(String department) { this.department = department; }
    public void setTeam(String team) { this.team = team; }
    public void setHeightCm(double heightCm) { this.heightCm = heightCm; }
    public void setWeightKg(double weightKg) { this.weightKg = weightKg; }
    public void setBmi(double bmi) { this.bmi = bmi; }
    public void setBloodGroup(String bloodGroup) { this.bloodGroup = bloodGroup; }
    public void setShiftType(String shiftType) { this.shiftType = shiftType; }
    public void setShiftStartTime(String shiftStartTime) { this.shiftStartTime = shiftStartTime; }
    public void setShiftEndTime(String shiftEndTime) { this.shiftEndTime = shiftEndTime; }
    public void setReminderWindow1(String reminderWindow1) { this.reminderWindow1 = reminderWindow1; }
    public void setReminderWindow2(String reminderWindow2) { this.reminderWindow2 = reminderWindow2; }
    public void setReminderFrequency(String reminderFrequency) { this.reminderFrequency = reminderFrequency; }
    public void setRole(String role) { this.role = role; }
    public void setStatus(String status) { this.status = status; }
    public void setConsentAccepted(boolean consentAccepted) { this.consentAccepted = consentAccepted; }
    public void setConsentDate(String consentDate) { this.consentDate = consentDate; }
    public void setOnboardingComplete(boolean onboardingComplete) { this.onboardingComplete = onboardingComplete; }
    public void setCurrentStreak(int currentStreak) { this.currentStreak = currentStreak; }
    public void setLongestStreak(int longestStreak) { this.longestStreak = longestStreak; }
    public void setCreatedAt(String createdAt) { this.createdAt = createdAt; }
    public void setUpdatedAt(String updatedAt) { this.updatedAt = updatedAt; }

    /**
     * Computes BMI from stored height and weight if not already set by server.
     */
    public double computeBmi() {
        if (heightCm > 0 && weightKg > 0) {
            double heightM = heightCm / 100.0;
            return weightKg / (heightM * heightM);
        }
        return 0;
    }
}

File: ios/BluePondWellness/Models/Participant.swift
// Participant.swift
// BluePond Wellness

import Foundation

struct Participant: Codable {
    var id: String?
    var authUserId: String?
    var displayName: String
    var employeeId: String
    var email: String
    var department: String?
    var team: String?
    var heightCm: Double?
    var weightKg: Double?
    var bmi: Double?
    var bloodGroup: String?
    var shiftType: String
    var shiftStartTime: String?
    var shiftEndTime: String?
    var reminderWindow1: String?
    var reminderWindow2: String?
    var reminderFrequency: String
    var role: String
    var status: String
    var consentAccepted: Bool
    var consentDate: String?
    var onboardingComplete: Bool
    var currentStreak: Int
    var longestStreak: Int
    var createdAt: String?
    var updatedAt: String?

    enum CodingKeys: String, CodingKey {
        case id
        case authUserId         = "auth_user_id"
        case displayName        = "display_name"
        case employeeId         = "employee_id"
        case email
        case department
        case team
        case heightCm           = "height_cm"
        case weightKg           = "weight_kg"
        case bmi
        case bloodGroup         = "blood_group"
        case shiftType          = "shift_type"
        case shiftStartTime     = "shift_start_time"
        case shiftEndTime       = "shift_end_time"
        case reminderWindow1    = "reminder_window_1"
        case reminderWindow2    = "reminder_window_2"
        case reminderFrequency  = "reminder_frequency"
        case role
        case status
        case consentAccepted    = "consent_accepted"
        case consentDate        = "consent_date"
        case onboardingComplete = "onboarding_complete"
        case currentStreak      = "current_streak"
        case longestStreak      = "longest_streak"
        case createdAt          = "created_at"
        case updatedAt          = "updated_at"
    }

    // MARK: - Computed

    var bmiCategory: String {
        guard let bmiValue = bmi else { return "Unknown" }
        switch bmiValue {
        case ..<18.5:  return "Underweight"
        case 18.5..<25: return "Healthy"
        case 25..<30:  return "Overweight"
        default:       return "Obese"
        }
    }

    var displayInitials: String {
        let parts = displayName.split(separator: " ").map(String.init)
        let first = parts.first?.first.map(String.init) ?? ""
        let last  = parts.count > 1 ? (parts.last?.first.map(String.init) ?? "") : ""
        return (first + last).uppercased()
    }
}
